M365/MDO/ResendFailedMail.ps1

```diff
@@ -0,0 +1,322 @@
+        if ($recipientCache[$failedMessage.SenderAddress]) {
+            $verifiedAcceptedSenderMessages += $failedMessage
+        } else {
+            Write-Verbose "Sender $($failedMessage.SenderAddress) is not a recipient in this tenant."
+            Write-Verbose "Discarded $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+        }
+    } else {
+        try {
+            $tempAddress = Get-EXOMailbox $failedMessage.SenderAddress -ErrorAction Stop
+            if ($null -eq $tempAddress) {
+                Write-Verbose "Sender $($failedMessage.SenderAddress) is not a recipient in this tenant."
+                Write-Verbose "Discarded $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+                $recipientCache[$failedMessage.SenderAddress] = $false
+            } else {
+                Write-Verbose "Added to cache Recipient $($failedMessage.SenderAddress) with Id $($failedMessage.SenderAddress)"
+                $recipientCache[$failedMessage.SenderAddress] = $true
+                Write-Verbose "Verified $($failedMessage.SenderAddress)"
+                $verifiedAcceptedSenderMessages += $failedMessage
+            }
+        } catch {
+            Write-Verbose "Error getting Sender Address $($failedMessage.SenderAddress)"
+            Write-Verbose "Discarded $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+            $recipientCache[$failedMessage.SenderAddress] = $false
+            Write-Verbose "$_"
+        }
+    }
+}
+
+$totalMessages = $verifiedAcceptedSenderMessages.Count
+if ($totalMessages -gt 0) {
+    if (-not $Force) {
+        Write-Host "`nWe are going to resend the following messages:"
+        Write-Host ($verifiedAcceptedSenderMessages | Format-Table -AutoSize Received, MessageId, SenderAddress, RecipientAddress, Subject | Out-String)
+        Write-Host "Total number of messages: $totalMessages`n"
+    }
+
+    if ($Force -or $PSCmdlet.ShouldContinue("Are you sure you want to do it?", "Resend messages")) {
+        $count = 0
+        $resendCount = 0
+        foreach ( $failedMessage in $verifiedAcceptedSenderMessages ) {
+            $count++
+            Write-Progress -Activity "Resending Progress" -Status "$count of $totalMessages" -PercentComplete ($count / $totalMessages * 100) -CurrentOperation "Resending message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)"
+            try {
+                $fullMessage = $null
+                $fullMessage = Get-MgUserMessage -UserId $failedMessage.SenderAddress -Filter "InternetMessageId eq '$($failedMessage.MessageId)'" -ExpandProperty Attachments -ErrorAction Stop | Sort-Object ReceivedDateTime | Select-Object -First 1
+            } catch {
+                Write-Host "Error getting message $($failedMessage.MessageId) - Subject: $($failedMessage.Subject)" -ForegroundColor Red
+                Write-Verbose "$_"
```

Comment:
We should add a `continue` here, as we failed to get the message. Move onto the next message in the loop.

Reply:
Not required, but something to think about to quickly move to the next item in the loop.